exchanges/okx/okx_websocket.go

```diff
@@ -0,0 +1,1858 @@
+	OkxChannelMarkPriceCandle3Dutc  = markPrice + OkxChannelCandle3Dutc
+	OkxChannelMarkPriceCandle5Dutc  = markPrice + OkxChannelCandle5Dutc
+	OkxChannelMarkPriceCandle12Hutc = markPrice + OkxChannelCandle12Hutc
+	OkxChannelMarkPriceCandle6Hutc  = markPrice + OkxChannelCandle6Hutc
+)
+
+// WsConnect initiates a websocket connection
+func (ok *Okx) WsConnect() error {
+	if !ok.Websocket.IsEnabled() || !ok.IsEnabled() {
+		return errors.New(stream.WebsocketNotEnabled)
+	}
+	var dialer websocket.Dialer
+	dialer.ReadBufferSize = 8192
+	dialer.WriteBufferSize = 8192
+	var authDialer websocket.Dialer
+	authDialer.ReadBufferSize = 8192
+	authDialer.WriteBufferSize = 8192
+	err := ok.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	if ok.Verbose {
+		log.Debugf(log.ExchangeSys, "Successful connection to %v\n",
+			ok.Websocket.GetWebsocketURL())
+	}
+	ok.Websocket.Conn.SetupPingHandler(stream.PingHandler{
+		UseGorillaHandler: true,
+		MessageType:       websocket.PingMessage,
+		Delay:             time.Second * 5,
+	})
+
+	ok.Websocket.SetCanUseAuthenticatedEndpoints(true)
+	if ok.IsWebsocketAuthenticationSupported() {
+		go func() {
+			er := ok.WsAuth(context.Background(), &authDialer)
+			if er != nil {
+				return
+			}
+			ok.Websocket.Wg.Add(1)
+			go ok.wsFunnelConnectionData(ok.Websocket.AuthConn)
+		}()
+	}
+	ok.Websocket.Wg.Add(2)
+	go ok.wsFunnelConnectionData(ok.Websocket.Conn)
+	go ok.WsReadData()
+	return nil
+}
+
+// WsAuth will connect to Okx's Private websocket connection and Authenticate with a login payload.
+func (ok *Okx) WsAuth(ctx context.Context, dialer *websocket.Dialer) error {
+	if !ok.Websocket.CanUseAuthenticatedEndpoints() {
+		return fmt.Errorf("%v AuthenticatedWebsocketAPISupport not enabled", ok.Name)
+	}
+	var creds *account.Credentials
+	err := ok.Websocket.AuthConn.Dial(dialer, http.Header{})
+	if err != nil {
+		return fmt.Errorf("%v Websocket connection %v error. Error %v", ok.Name, okxAPIWebsocketPrivateURL, err)
+	}
+	ok.Websocket.AuthConn.SetupPingHandler(stream.PingHandler{
+		UseGorillaHandler: true,
+		MessageType:       websocket.PingMessage,
+		Delay:             time.Second * 5,
+	})
+	creds, err = ok.GetCredentials(ctx)
+	if err != nil {
+		return err
+	}
+	ok.Websocket.SetCanUseAuthenticatedEndpoints(true)
+	timeUnix := time.Now()
+	signPath := "/users/self/verify"
+	hmac, err := crypto.GetHMAC(crypto.HashSHA256,
+		[]byte(strconv.FormatInt(timeUnix.UTC().Unix(), 10)+http.MethodGet+signPath),
+		[]byte(creds.Secret),
+	)
+	if err != nil {
+		return err
+	}
+	base64Sign := crypto.Base64Encode(hmac)
+	request := WebsocketEventRequest{
+		Operation: "login",
+		Arguments: []WebsocketLoginData{
+			{
+				APIKey:     creds.Key,
+				Passphrase: creds.ClientID,
+				Timestamp:  timeUnix,
+				Sign:       base64Sign,
+			},
+		},
+	}
+	go func() {
+		var response []byte
+		response, err = ok.Websocket.AuthConn.SendMessageReturnResponse("login", request)
```

Comment:
Since the login response is coming late, I am handling the response in the WsHandleData() function. you can check it out in okx_websocket.go L513.
Let me try another implementation alternative and get back to u